### PR TITLE
Properly detect removable variables

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/6231

Depends on #6118

### CODE REVIEW

**Comment by erak** [comment]
Will review after https://github.com/ethereum/solidity/pull/6118 was merged :)

File: libyul/optimiser/StackCompressor.cpp [hunk 1]
```diff
@@ -74,15 +74,15 @@ void eliminateVariables(shared_ptr<Dialect> const& _dialect, ASTNode& _node, siz
...
 
-bool StackCompressor::run(shared_ptr<Dialect> const& _dialect, Block& _ast)
+bool StackCompressor::run(shared_ptr<Dialect> const& _dialect, Block& _ast, bool _optimizeStackAllocation)
 {
 	yulAssert(
 		_ast.statements.size() > 0 && _ast.statements.at(0).type() == typeid(Block),
 		"Need to run the function grouper before the stack compressor."
 	);
-	for (size_t iterations = 0; iterations < 4; iterations++)
+	for (size_t iterations = 0; iterations < 6; iterations++)
```
Comment (Marenz) [inline libyul/optimiser/StackCompressor.cpp]: Why the iteration change?

Comment (chriseth) [reply libyul/optimiser/StackCompressor.cpp]: 4 turned out to be a little low for real-world examples. Also, we have the `return` below, so we can even set it higher.

**Review by Marenz — DISMISSED** [review]
Code seems good to me. Didn't completely follow the Rematerializer Optimization. 
I personally would put the number change into a different commit as it is unrelated to "passing down flags" as the commit says.
Maybe also make it a const/define to prevent magic numbers.

**Comment by leonardoalt** [comment]
Compilation is failing. Does it need to be rebased?

File: libyul/optimiser/StackCompressor.cpp [hunk 2]
```diff
@@ -39,24 +39,86 @@ using namespace yul;
...
+	/// Number of references to each candidate variable.
+	map<YulString, size_t> m_numReferences;
+};
+
+template <typename ASTNode>
+void eliminateVariables(shared_ptr<Dialect> const& _dialect, ASTNode& _node, size_t _numVariables)
+{
 	set<pair<size_t, YulString>> rematCosts;
-	for (auto const& ssa: ssaValues.values())
 	{
```
Comment (leonardoalt) [inline libyul/optimiser/StackCompressor.cpp]: Why keep these?

**Review by leonardoalt — APPROVED** [review]
Lgtm

**Review by leonardoalt — CHANGES_REQUESTED** [review]
Oh wait, forgot that compilation is still failing.

File: libyul/optimiser/StackCompressor.cpp (hunk 2 above)
Comment (chriseth) [reply libyul/optimiser/StackCompressor.cpp]: Will simplify that.

**Comment by chriseth** [comment]
Updated.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**